### PR TITLE
[OS Data Sync] Google fit permissions

### PR DESCRIPTION
**Pull request by brendeny**
This PR integrates Google fit into the app. The onboarding screen should ask for google fit permissions and help the user set permissions to gather the google fit data. There should be a homescreen prompt if the user does not grant permissions.

## Changes
- The changes here implement the react-native-google-fit npm package for google fit auth. Users are able to be added on firebase as a test user and their google fit data can be accessed from their accounts once authed in the onboarding screen. Additionally, auth flows are introduced on the homescreen and settings screen.

## Requirements
N/A

Requires: 
Users have installed Fit App and have Steps and Sleep data

## PR Dependencies
- Depends on #52

## Merged Branches and Other Dependencies
Merged branch(es): https://github.com/bipolarbridges/companion-app/pull/52 

## Subtasks / Progress
- [x] Auth users on onboarding
- [x] Fetch Sleep and Step data
- [x] Card on homescreen if user skips onboarding
- [x] Auth through card on homescreen if user skips onboarding
- [x] Settings page auth

## Preview
This is the google-fit auth flow for onboarding:
https://user-images.githubusercontent.com/46551861/144527268-0a3ecd12-d3e3-4707-9ac0-b9adb6906855.mp4

This is the homescreen google fit auth flow when the user skips auth in the onboarding:
https://user-images.githubusercontent.com/46551861/144527323-dcc6866a-791c-4a2e-ad1f-7b98a8fbc040.mp4

This is the google fit auth through user settings:
https://user-images.githubusercontent.com/46551861/144527347-d38d85d4-fa35-4b99-8291-9d95b8c56ae9.mp4

## Testing
- [x] Manual 
- [ ] Automated

List of added tests:
- N/A

### CODE REVIEW

File: mobile/src/stateMachine/views/healthData/healthScopes.tsx [hunk 1]
```diff
@@ -73,6 +75,27 @@ export class HealthScopesView extends ViewState {
...
+            customView:
+                <View>
+                    <View style={styles.imageNextToText}>
+                        <Image source={Images.GoogleFitHealthPermPNG} style={{ width: 26, height: 26 }} />
+                        <Text style={[TextStyles.p1, styles.instructionText]}>{'Turn All Categories On'}</Text>
+                    </View>
+                </View>,
+            onConfirm: async () => {
+                AppController.Instance.PromptModal.closeModal();
+                await authAndroid();
```
Comment (brendeny) [inline mobile/src/stateMachine/views/healthData/healthScopes.tsx:95]: @MarkDavies0 there is a small fix needed here where the screen the user ends on after pressing confirm needs to refresh to show the updated auth. Or another option is to send the user to the homescreen after auth success

**Comment by brendeny** [comment]
I have implemented the Google Fit auth and tested it. The permissions are persistent in the users google account once the user grants them, however currently, the user needs to auth every time they close/open the app. This may be because every time the app is refreshed as it is in a development/testing version

File: mobile/src/stateMachine/views/healthData/healthScopes.tsx (hunk 1 above)
Comment (MarkDavies0) [reply mobile/src/stateMachine/views/healthData/healthScopes.tsx:95]: @brendeny I think you need to `await authAndroid()` before calling `AppController.Instance.PromptModal.closeModal();` . Alternatively you could have the view react to the new auth status and re-render itself

File: mobile/src/constants/images/app.ts [hunk 2]
```diff
@@ -49,6 +49,7 @@ import GoalsSVG from 'src/assets/images/app/goals-icon.svg';
 import GoalsActiveSVG from 'src/assets/images/app/goals-icon-active.svg';
 import GoalsIndicatorSVG from 'src/assets/images/app/goals-icon-indicator.svg';
 import GoalsIndicatorActiveSVG from 'src/assets/images/app/goals-icon-indicator-active.svg';
+import GoogleFitHealthPermPNG from 'src/assets/images/android_check.png'
```
Comment (MarkDavies0) [inline mobile/src/constants/images/app.ts:52]: Naming here is off, the actual file google_fit_health_perm.png in unused while android_check.png is being imported as GoogleFitHealthPermPNG

File: mobile/src/helpers/health.ts [hunk 3]
```diff
@@ -2,14 +2,138 @@ import GoogleFit, { Scopes } from 'react-native-google-fit';
...
+    }
+  } else {
+    logger.log('Daily Steps Not Found');
+  }
+};
+
+const fetchSleepData = async opt => {
+  const midnight = "00:00"
+  // var midnight = new Date();
+  // midnight.setHours(0, 0, 0, 0);
```
Comment (MarkDavies0) [inline mobile/src/helpers/health.ts:48]: removed commented code throughout

File: mobile/src/helpers/health.ts [hunk 4]
```diff
@@ -2,14 +2,138 @@ import GoogleFit, { Scopes } from 'react-native-google-fit';
...
+  // setSleep(Math.round((sleepTotal / (1000 * 60 * 60)) * 100) / 100);
+};
+
+const fetchAndroidHealthData = async () => {
+      let sleepData;
+      let stepsData;
+      // call Google Fit API to get the steps data for the user
+      // opts --> options object
+      logger.log('GOOGLE_FIT: Fetching Step Data');
+      stepsData = await fetchStepsData(opt);
```
Comment (MarkDavies0) [inline mobile/src/helpers/health.ts:79]: make declaration inline with assignment
`let stepsData = await fetchStepsData(opt);`

File: mobile/src/helpers/health.ts [hunk 5]
```diff
@@ -2,14 +2,138 @@ import GoogleFit, { Scopes } from 'react-native-google-fit';
...
-    Scopes.FITNESS_BODY_READ,
-  ]
+    Scopes.FITNESS_SLEEP_READ
+  ],
 };
 
+//Google Fit Steps
+let dailyStepCount;
+
+var today = new Date();
```
Comment (MarkDavies0) [inline mobile/src/helpers/health.ts:16]: prefer let or const over var, probably const here

File: mobile/src/helpers/health.ts [hunk 6]
```diff
@@ -2,14 +2,138 @@ import GoogleFit, { Scopes } from 'react-native-google-fit';
...
+  endDate: today.toISOString(), // required ISO8601Timestamp
+  bucketUnit: 'DAY', // optional - default "DAY". Valid values: "NANOSECOND" | "MICROSECOND" | "MILLISECOND" | "SECOND" | "MINUTE" | "HOUR" | "DAY"
+  bucketInterval: 1, // optional - default 1.
+};
+
+const fetchStepsData = async opt => {
+  const res = await GoogleFit.getDailyStepCountSamples(opt);
+  if (res.length !== 0) {
+    for (var i = 0; i < res.length; i++) {
+      if (res[i].source === 'com.google.android.gms:estimated_steps') {
```
Comment (MarkDavies0) [inline mobile/src/helpers/health.ts:33]: This for loop seems like overkill, maybe try Array.find

File: mobile/src/helpers/health.ts [hunk 7]
```diff
@@ -2,14 +2,138 @@ import GoogleFit, { Scopes } from 'react-native-google-fit';
...
+      logger.log('GOOGLE_FIT: Fetching Step Data');
+      stepsData = await fetchStepsData(opt);
+
+      // call Google Fit API to get the sleep data for the user
+      // opts --> options object
+      logger.log('GOOGLE_FIT: Fetching Sleep Data');
+      sleepData = await fetchSleepData(opt);
+
+      logger.log('GOOGLE_FIT: Sleep Data: ', JSON.stringify(sleepData));
+      logger.log('GOOGLE_FIT: Step Data: ', JSON.stringify(stepsData));
```
Comment (MarkDavies0) [inline mobile/src/helpers/health.ts:87]: There are a lot of logger and console.log statements in this file, remove all except the most useful for future devs

File: mobile/src/helpers/health.ts [hunk 8]
```diff
@@ -2,14 +2,138 @@ import GoogleFit, { Scopes } from 'react-native-google-fit';
...
     Scopes.FITNESS_ACTIVITY_READ,
-    Scopes.FITNESS_ACTIVITY_WRITE,
-    Scopes.FITNESS_BODY_READ,
-  ]
+    Scopes.FITNESS_SLEEP_READ
+  ],
 };
 
+//Google Fit Steps
+let dailyStepCount;
```
Comment (MarkDavies0) [inline mobile/src/helpers/health.ts:14]: Move this into the scope where it is used

File: mobile/src/helpers/health.ts (hunk 5 above)
Comment (MarkDavies0) [reply mobile/src/helpers/health.ts:16]: `lastWeekDate` should be const

File: mobile/src/helpers/health.ts [hunk 9]
```diff
@@ -2,14 +2,138 @@ import GoogleFit, { Scopes } from 'react-native-google-fit';
...
+        return dailyStepCount;
+      }
+    }
+  } else {
+    logger.log('Daily Steps Not Found');
+  }
+};
+
+const fetchSleepData = async opt => {
+  const midnight = "00:00"
```
Comment (MarkDavies0) [inline mobile/src/helpers/health.ts:46]: `midnight` is used as a date object so I would declare it as an actual date and save on Date.parse further down

File: mobile/src/helpers/health.ts [hunk 10]
```diff
@@ -2,14 +2,138 @@ import GoogleFit, { Scopes } from 'react-native-google-fit';
...
+  // var midnight = new Date();
+  // midnight.setHours(0, 0, 0, 0);
+  let sleepTotal = 0;
+  const res = await GoogleFit.getSleepSamples(opt);
+
+  for (var i = 0; i < res.length; i++) {
```
Comment (MarkDavies0) [inline mobile/src/helpers/health.ts:52]: this is for loop is quite difficult to interpret, any way to simplify? A comment describing what the loop is doing would be helpful

File: mobile/src/stateMachine/scenario.ts [hunk 11]
```diff
@@ -185,7 +185,8 @@ export const MasloScenario: GlobalScenario<States> = {
             { priority: 0, target: States.Consent, condition: VM.showConsent },
             { priority: 1, target: States.OnboardingEnter, condition: VM.hasActiveOnboarding },
             { priority: 2, target: States.AskNotificationsPermissions, condition: VM.askNotifyPermissions },
-            { priority: 3, target: States.HealthConsent, condition: Platform.OS == 'ios' ? VM.needsHealthPromptIOS : VM.hasHealthPermissions },
+            { priority: 3, target: States.HealthConsent, condition: Platform.OS == 'ios' ? VM.needsHealthPromptIOS : VM.needsHealthPromptAndroid },
+            // { priority: 3, target: States.HealthConsent, condition: Platform.OS == 'android' ? VM.needsHealthPromptAndroid : VM.hasHealthPermissions },
```
Comment (MarkDavies0) [inline mobile/src/stateMachine/scenario.ts:189]: remove commented line

File: mobile/src/stateMachine/views/main/home.tsx [hunk 12]
```diff
@@ -293,9 +296,18 @@ export class HomeView extends ViewState<{ opacity: Animated.Value, isUnfinishedQ
...
 
     private showHealthPermissionCard() {
-        return Platform.OS == 'ios' && !this.healthPermissionsEnabled;
+        return Platform.OS == 'ios'? !this.healthPermissionsEnabled : !checkAndroidAuth();
     }
 
+    // private healthPermAnd() {
+    //     this.authAn();
+    //     return checkAndroidAuth();
+    // }
```
Comment (MarkDavies0) [inline mobile/src/stateMachine/views/main/home.tsx:305]: remove

File: mobile/src/stateMachine/views/main/home.tsx [hunk 13]
```diff
@@ -55,10 +58,10 @@ export class HomeView extends ViewState<{ opacity: Animated.Value, isUnfinishedQ
     get viewQolModel() { return AppViewModel.Instance.QOL; }
 
     async start() {
-        await AppViewModel.Instance.QOL.init();
-        const qolArmMagnitudes = await this.viewModel.getArmMagnitudes();
-        this.persona.qolArmMagnitudes = qolArmMagnitudes;
-        this.setState({ ...this.state, isUnfinishedQol: AppViewModel.Instance.QOL.isUnfinished });
+        // await AppViewModel.Instance.QOL.init();
+        // const qolArmMagnitudes = await this.viewModel.getArmMagnitudes();
```
Comment (MarkDavies0) [inline mobile/src/stateMachine/views/main/home.tsx:62]: uncomment right before merge

**Review by MarkDavies0 — CHANGES_REQUESTED** [review]
Some changes requested but looking good. Also make sure to delete the template parts of the PR template

File: mobile/src/helpers/health.ts (hunk 9 above)
Comment (brendeny) [reply mobile/src/helpers/health.ts:46]: I had tried to set midnight as a date object, however in order to make the comparison between re[i].startDate and midnight, we need it as a string.

let midnight = new Date();
midnight.setHours(0,0,0,0);

File: mobile/src/helpers/health.ts (hunk 10 above)
Comment (brendeny) [reply mobile/src/helpers/health.ts:52]: left comment

**Comment by brendeny** [comment]
@MarkDavies0 thanks for the review! Made the changes requested. Feel free to reach out to me if you have any further questions. Cheers

**Comment by MarkDavies0** [comment]
@wynnset Tested on iOS and Android devices, needs final review before merge

File: mobile/src/stateMachine/scenario.ts [hunk 14]
```diff
@@ -192,7 +192,7 @@ export const MasloScenario: GlobalScenario<States> = {
             { priority: 0, target: States.Consent, condition: VM.showConsent },
             { priority: 1, target: States.OnboardingEnter, condition: VM.hasActiveOnboarding },
             { priority: 2, target: States.AskNotificationsPermissions, condition: VM.askNotifyPermissions },
-            { priority: 3, target: States.HealthConsent, condition: Platform.OS == 'ios' ? VM.needsHealthPromptIOS : VM.hasHealthPermissions },
```
Comment (wynnset) [inline mobile/src/stateMachine/scenario.ts:195]: How come we are no longer checking `VM.hasHealthPermissions`?

**Review by wynnset — COMMENTED** [review]
Hey @MarkDavies0 it seems like the file "mobile/src/assets/images/google_fit_health_perm.png" is not used anywhere. Or am I mistaken? Also, please check the one comment below.

File: mobile/src/stateMachine/scenario.ts (hunk 14 above)
Comment (MarkDavies0) [reply mobile/src/stateMachine/scenario.ts:195]: The new methods for android and iOS are used instead

**Review by wynnset — APPROVED** [review]

**Pull request merged**